### DISCUSSION TITLE
How to add css-in-js library when using a Vike extension

### DISCUSSION THREAD

**Discussion by edmelly-pg** [vikejs/vike#1486] [Discussion] [answered]
I want to make use of `vike-react`, but I also wish to set up `styled-components`. If `vike-react` is handling the `onRenderHtml()` hook ([here](https://github.com/vikejs/vike-react/blob/main/packages/vike-react/src/renderer/onRenderHtml.tsx)), is there any way of collecting styles (with, e.g., `ServerStyleSheet`) and injecting them, or do I have to switch to manual integration?

Thanks!

**Comment by brillout** [comment, marked as answer]
The plan is to have `vike-react` provide hooks enabling such integration. We can then also implement `vike-react-styled-components`.

Let me know if you'd be up for a PR; the `vike-react` source code is quite small & digestible. If not then you'll have to wait a little (maybe like 1-2 weeks) until I'm done with I'm the high priorities I'm currently working on.

**Reply by edmelly-pg** [reply]
Thanks for your quick response. I managed to add `styled-components` to `vike-react` locally, but:
- It's hard-coded (i.e. import and logic added to `onRenderHtml`). I am struggling to see how to make this work as a new hook/how to allow for a generic CSS-in-JS approach to cover multiple libraries.

![image](https://github.com/vikejs/vike/assets/90700946/4c9fef51-0205-475a-b521-fd69b1062de5)

- Only when `config.stream` is `false`. If `stream` is true, then `renderToStream` from `react-streaming/server` is used instead of `renderToString`, so I think implementing the [streaming rendering API](https://styled-components.com/docs/advanced#streaming-rendering) with `sheet.interleaveWithNodeStream()` needs to happen there.

**Reply by brillout** [reply]
I think, but correct me if I'm wrong, a solution would be that `vike-react` creates two new custom hooks: `on{Before,After}RenderHtml()`. Then the user-land and/or `vike-style-components` would use these two hooks to implement the integration.

**Reply by Pixelatex** [reply]
@brillout has this fix landed yet? Currently attempting to get styled-comp working with vike-react

**Reply by brillout** [reply]
Yes

**Reply by momaqbol** [reply]
@brillout 
I can't find that "on{Before,After}RenderHtml()" hook anywhere, can you please point me to the release of this fix?

**Reply by brillout** [reply]
https://vike.dev/onAfterRenderHtml

**Reply by brillout** [reply]
Ah, actually `vike-react` doesn't implement it (see the list at the bottom of https://vike.dev/vike-react#under-the-hood). PR welcome to add it.
